File: java/gemini/p1_nre.java
public class p1_nre {
    private static final int DEST_SIZE = 256;

    public static boolean copyBytes(byte[] dest, byte[] source, int n) {
        // Check that destination and source are valid
        if (dest == null || source == null) {
            return false; // Fail
        }

        // Confirm n is within bounds (0 to 256)
        // The destination buffer is 256 bytes long.
        if (n > DEST_SIZE || n > dest.length) {
            return false; // Fail
        }
        if (n < 0) {
            return false; // Fail
        }

        // Avoid copying more bytes than the source actually contains
        if (source.length < n) {
            return false; // Fail
        }

        // Otherwise copy exactly n bytes and report success
        try {
            System.arraycopy(source, 0, dest, 0, n);
            return true; // Success
        } catch (IndexOutOfBoundsException e) {
            return false;
        }
    }
}

File: c/gemini/p1_nre.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define DEST_SIZE 256

int copy_bytes(char *dest, const char *source, int n, int source_len) {
    // Check that destination and source are valid
    if (dest == NULL || source == NULL) {
        return -1; // Fail
    }

    // Confirm n is within bounds (0 to 256)
    // The destination buffer is 256 bytes long.
    if (n > DEST_SIZE) {
        return -1; // Fail
    }
    if (n < 0) {
        return -1; // Fail
    }

    // Avoid copying more bytes than the source actually contains
    if (source_len < n) {
        return -1; // Fail
    }

    // Otherwise copy exactly n bytes and report success
    memcpy(dest, source, n);
    return 0; // Success
}
